flink/v1.14/flink/src/main/java/org/apache/iceberg/flink/sink/EqualityFieldKeySelector.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.sink;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.lang.reflect.Array;
+import java.nio.ByteBuffer;
+import java.util.List;
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.iceberg.Accessor;
+import org.apache.iceberg.Schema;
+import org.apache.iceberg.StructLike;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.types.Conversions;
+import org.apache.iceberg.util.ByteBuffers;
+
+/**
+ * Create a {@link KeySelector} to shuffle by equality fields, to ensure same equality fields record will be emitted to
+ * same writer. That can prevent create duplicate record when insert and delete one row which have same equality field
+ * values on different writer in one transaction, and guarantee pos-delete will take effect.
+ */
+class EqualityFieldKeySelector extends BaseKeySelector<RowData, ByteBuffer> {
+
+  private final Accessor<StructLike>[] accessors;
+
+  @SuppressWarnings("unchecked")
+  EqualityFieldKeySelector(List<Integer> equalityFieldIds, Schema schema, RowType flinkSchema) {
+    super(schema, flinkSchema);
+
+    int size = equalityFieldIds.size();
+    this.accessors = (Accessor<StructLike>[]) Array.newInstance(Accessor.class, size);
+    for (int i = 0; i < size; i++) {
+      Accessor<StructLike> accessor = schema.accessorForField(equalityFieldIds.get(i));
+      Preconditions.checkArgument(accessor != null,
+          "Cannot build accessor for field: {}", schema.findField(equalityFieldIds.get(i)));
+
+      accessors[i] = accessor;
+    }
+  }
+
+  @Override
+  public ByteBuffer getKey(RowData row) throws Exception {
```

Comment:
this is used in the hot code path of per-record processing. small concern on the number of ByteBuffer creation/wrap, concatenation. but not sure how to make it better and maybe we don't pre-maturely optimize it until is proved to be a problem.

Reply:
Agree!

Reply:
We don't need to construct a totally new `ByteBuffer` for generating the shuffling key.  Actually we can only construct a lightweight `StructLikeWrapper` which will calculate its hashCode for shuffling purpose.   Because in the root path,  the `KeyGroupRangeAssignment` only need the key's `hashCode`. see https://github.com/apache/flink/blob/master/flink-runtime/src/main/java/org/apache/flink/runtime/state/KeyGroupRangeAssignment.java#L65 . 

As this is a hot code path as @stevenzwu said, I will strongly suggest to use the shadow clone approach rather then the deep clone one.

Reply:
Nice catch! I have rolled back the implementation of `EqualityFieldKeySelector` to use `StructLikeWrapper` and return the `hashCode` of `Struct Like Wrapper`. Please read it again!